File: leetcode/236-二叉树的最近公共祖先/236-二叉树的最近公共祖先.js
/**
 * Definition for a binary tree node.
 */
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */

// 236-二叉树的最近公共祖先
var lowestCommonAncestor = function (root, p, q) {
    if (!root) {
        return null;
    }
    if (root === p || root === q) {
        return root;
    }
    const left = lowestCommonAncestor(root.left, p, q)
    const right = lowestCommonAncestor(root.right, p, q)
    if (!left && !right) {
        return null;
    }
    if (left && right) {
        return root;
    }
    return left === null ? right : left
};

File: leetcode/236-二叉树的最近公共祖先/236-二叉树的最近公共祖先.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 236-二叉树的最近公共祖先
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None:
            return None
        if p == root or q == root:
            return root
        left = self.lowestCommonAncestor(root.left,p,q)
        right = self.lowestCommonAncestor(root.right,p,q)
        if not left is None and not right is None:
            return root
        if left is None and right is None:
            return None

        if left is None:
            return right 
        else:
            return left